src/sentry/static/sentry/app/actionCreators/globalSelection.tsx

```diff
@@ -63,9 +77,137 @@ export function resetGlobalSelection() {
   GlobalSelectionActions.reset();
 }
 
+function getProjectIdFromProject(project: MinimalProject) {
+  return parseInt(project.id, 10);
+}
+
+type InitializeUrlStateParams = {
+  organization: Organization;
+  queryParams: ReactRouter.WithRouterProps['location']['query'];
+  router: ReactRouter.WithRouterProps['router'];
+  memberProjects: Project[];
+  shouldForceProject?: boolean;
+  shouldEnforceSingleProject: boolean;
+  /**
+   * If true, do not load from local storage
+   */
+  skipLastUsed?: boolean;
+  defaultSelection?: Partial<GlobalSelection>;
+  forceProject?: MinimalProject | null;
+};
+
+export function initializeUrlState({
```

Comment:
This logic was spread out among the store and the component, I've moved them all in here so that it's easier to reason about all of these state changes.